docs: add parameterizability guide to our docs (#464)
kurtosis-tech/kurtosis: docs/docs/guides/how-to-parameterize-cassandra.md

---
title: How to set up an n-node Cassandra environment
sidebar_label: Setting up an n-node Cassandra cluster
slug: /how-to-parameterize-cassandra
toc_max_heading_level: 2
sidebar_position: 5
---

Introduction
------------

> In this guide, we will set up a 3-node Cassandra cluster in Docker and parameterize the environment definition so it can easily be modified for use in different tests that require an _n_-node Cassandra cluster. Then we will show you how to run remotely hosted packages authored by others, and go through how to package and publish our work to Github for others to use as well.

Reviewer: ```suggestion
In this guide, you will set up a 3-node Cassandra cluster in Docker and parameterize the environment definition so it can easily be modified for use in different tests that require an _n_-node Cassandra cluster. Then we will show you how to run remotely hosted packages authored by others, and go through how to package and publish our work to Github for others to use as well.
```

> Specifically, we're going to configure our test environments with a way that allows us to both:

Reviewer: ```suggestion
Specifically, you're going to configure your test environments with a way that allows you to both:
```

> 1. Parameterize the environment so a user can trivially specify how many nodes they’d like for their system to have, and 

Reviewer: ```suggestion
1. Parameterize the environment so another developer using the environment can specify how many nodes they’d like for their system to have, and 
```

> 2. Make their environment definition composable so that these environments can be easily included in tests with other services for different scenarios & use cases.

Reviewer: ```suggestion
2. Make the environment definition composable so that your environments can be included in tests with other services for different scenarios & use cases.
```

**Without Kurtosis**

One way to accomplish the above would be to write shell scripts over docker, or over binaries running on bare metal. In this case, we’d have to build these capabilities into our scripts and handle cross-system issues (laptop vs CI) ourselves.

**With Kurtosis**

> However, in this guide, we’re going to use a free tool that already has these capabilities built in - Kurtosis. Kurtosis is a composable build system for writing reproducible test environments, and can run on your own laptop or in your favorite CI provider.

Reviewer: ```suggestion
In this guide, we’re going to use Kurtosis. Kurtosis has composability, parameterizability, and portability built into its environment definition system and runtime. With Kurtosis we can ensure that these environments are runnable on your own laptop or in your favorite CI provider.
```

:::info

> For a holistic view of Kurtosis’ capabilities, visit our [Quickstart][quickstart]!

Reviewer: I'd suggest just removing this info box, they have it in the sidebar

Author: Done

:::

Setup
-----
Before you proceed, make sure you have:
* [Installed and started the Docker engine on your local machine][starting-docker]
* [Installed the Kurtosis CLI (or upgraded it to the latest release, if you already have the CLI installed)][installing-the-cli]

Instantiate a 3-node Cassandra cluster
--------------------------------------

> First, create and `cd` into a directory to hold the project we’ll be working on:

Reviewer: ```suggestion
First, create and `cd` into a directory to hold the project you'll be working on:
```

    plan.wait(
        service_name = get_first_node_name(),
        recipe = check_nodes_are_up, 
        field = "output", 
        assertion = "==", 
        target_value = str(num_nodes), 
        timeout = "8m", 
        )

    return {"node_names": [get_service_name(x) for x in range(num_nodes)]}

def get_service_name(node_idx):
    return CASSANDRA_NODE_PREFIX + str(node_idx)

def get_service_config(num_nodes):
    seeds = ["cassandra-node-"+str(x) for x in range(0, num_nodes)]
    return ServiceConfig(
        image = CASSANDRA_NODE_IMAGE,
        ports = {
            CLUSTER_COMM_PORT_ID : PortSpec(number = CLUSTER_COM_PORT_NUMBER, transport_protocol = CLUSTER_COM_PORT_PROTOCOL),
            CLIENT_COMM_PORT_ID : PortSpec(number = CLIENT_COM_PORT_NUMBER, transport_protocol = CLIENT_COM_PORT_PROTOCOL),
        },
        env_vars = {
            "CASSANDRA_SEEDS":",".join(seeds),
            # without this set Cassandra tries to take 8G and OOMs
            "MAX_HEAP_SIZE": "512M",
            "HEAP_NEWSIZE": "1M",
        }
    )

def get_first_node_name():
    return get_service_name(FIRST_NODE_INDEX)
```

> Finally, save your newly created file and, from the working directory we created, run the following command:

Reviewer: ```suggestion
Finally, save your newly created file and, from the working directory you created, run the following command:
```

```bash
kurtosis run --enclave cassandra-cluster main.star
```

:::info

> Kurtosis will run validation checks against our code to ensure that it will work before spinning up the containers for our 3-node Cassandra cluster. We won’t dive into the details of how validation checks are used by Kurtosis in this guide, but you can read more about them [here][multi-phase-runs].

Reviewer: ```suggestion
Kurtosis will run validation checks against your code to ensure that it will work before spinning up the containers for our 3-node Cassandra cluster. We won’t dive into the details of how validation checks are used by Kurtosis in this guide, but you can read more about them [here][multi-phase-runs].
```

Wait took 33 tries (52.05875544s in total). Assertion passed with following:
Command returned with exit code '0' and the following output: 3

Starlark code successfully run. Output was:
{
	"node_names": [
		"cassandra-node-0",
		"cassandra-node-1",
		"cassandra-node-2"
	]
}
Name:            cassandra-cluster
UUID:            c8027468561c
Status:          RUNNING
Creation Time:   Tue, 28 Mar 2023 16:29:54 -03

========================================= Files Artifacts =========================================
UUID   Name

========================================== User Services ==========================================
UUID           Name               Ports                                  Status
ec084228aa2b   cassandra-node-0   client: 9042/tcp -> 127.0.0.1:52503    RUNNING
                                  cluster: 7000/tcp -> 127.0.0.1:52502
f605cff291ef   cassandra-node-1   client: 9042/tcp -> 127.0.0.1:52508    RUNNING
                                  cluster: 7000/tcp -> 127.0.0.1:52507
4bcf767e82f5   cassandra-node-2   client: 9042/tcp -> 127.0.0.1:52513    RUNNING
                                  cluster: 7000/tcp -> 127.0.0.1:52512
```

> Congratulations! We’ve used Kurtosis to spin up a three-node Cassandra cluster over Docker. 

Reviewer: ```suggestion
Congratulations! You’ve used Kurtosis to spin up a three-node Cassandra cluster over Docker. 
```

### Review

> In this section, we created file using [Starlark][starlark] that are instructions for Kurtosis to do the following:

Reviewer: ```suggestion
In this section, you created a [Starlark][starlark] file with instructions for Kurtosis to do the following:
```

1. Spin up 3 Cassandra containers (one for each node), 
2. Bootstrap each node to the cluster,
3. Map the default Cassandra node container ports to ephemeral local machine ports (described in their respective `ServiceConfig`), and
4. Verify using `nodetool` that the cluster is up, running, and has 3 nodes (just as we specified) before returning the names of our nodes.

> We now have a simple environment definition for Kurtosis to spin up a 3-node Cassandra cluster. This is useful for developers who need an easy way to reproduce and spin up a 3-node Cassandra cluster for testing. You may now be wondering: but what if I need more nodes?

Reviewer: ```suggestion
You now have a simple environment definition for Kurtosis to spin up a 3-node Cassandra cluster. You may now be wondering: but what if I need to change the number of nodes?
```

Fortunately, Kurtosis environment definitions can be parameterized. We’ll see just how easy it is to do so in the next section.

> Parameterize our Cassandra cluster

Reviewer: ```suggestion
Parameterize your Cassandra cluster
```

----------------------------------

> One of Kurtosis’s most powerful features is the out-of-the-box ability to parameterize environment definitions with very little overhead. This can be incredibly useful for developers who want to run different tests or scenarios against various configurations of their environment, without the burden of editing their Bash scripts or `docker-compose.yml` files repeatedly to modify the setup each time.

Reviewer: ```suggestion
Kurtosis enables users to parameterize environment definitions out-of-the-box. If you need to run tests against multiple different configurations of your environment, you'll be to do so without needing maintain different Bash scripts or `docker-compose.yml` files per test.
```

> We will parameterize our Cassandra cluster environment definition by adding only 2 lines of code to our `main.star` Starlark file. 

Reviewer: ```suggestion
You can parameterize our Cassandra cluster environment definition by adding 2 lines of code to your `main.star` Starlark file. 
```

Let’s add in those extra lines now. 

> In your `main.star` file, add in the following lines in the code block that describes the `plan` object:

Reviewer: ```suggestion
In your `main.star` file, add the following lines in the code block that describes the `plan` object:
```

```python
def run(plan, args):
	# Default number of Cassandra nodes in our cluster
	num_nodes = 3

	### <----------- NEW CODE -----------> ###
	if "num_nodes" in args:
		num_nodes = args["num_nodes"]
	### <----------- NEW CODE -----------> ###
	
	for node in range(0, num_nodes):
		node_name = get_service_name(node)
		config = get_service_config(num_nodes)
		plan.add_service(
            name = node_name,
			config = config,
        )

	# ...
```

> Now save your newly edited `main.star` file and run the following command to blow away our old enclave and spin up a new one with 5 nodes specified:

Reviewer: ```suggestion
Now save your newly edited `main.star` file and run the following command to blow away your old enclave and spin up a new one with 5 nodes:
```

Name:            cassandra-cluster
UUID:            c7985e32b076
Status:          RUNNING
Creation Time:   Tue, 28 Mar 2023 21:45:47 -03

========================================= Files Artifacts =========================================
UUID   Name

========================================== User Services ==========================================
UUID           Name               Ports                                  Status
9a9213d1d846   cassandra-node-0   client: 9042/tcp -> 127.0.0.1:54740    RUNNING
                                  cluster: 7000/tcp -> 127.0.0.1:54741
66c6907c6a8a   cassandra-node-1   client: 9042/tcp -> 127.0.0.1:54746    RUNNING
                                  cluster: 7000/tcp -> 127.0.0.1:54745
fe9be7991edd   cassandra-node-2   client: 9042/tcp -> 127.0.0.1:54761    RUNNING
                                  cluster: 7000/tcp -> 127.0.0.1:54760
daff154657ce   cassandra-node-3   client: 9042/tcp -> 127.0.0.1:54768    RUNNING
                                  cluster: 7000/tcp -> 127.0.0.1:54769
8ecd6c4b75a6   cassandra-node-4   client: 9042/tcp -> 127.0.0.1:54773    RUNNING
                                  cluster: 7000/tcp -> 127.0.0.1:54774
```

> Congratulations! We’ve just parameterized our Cassandra cluster environment definition and used it to instantiate a 5-node Cassandra cluster. You can run the same command with 2 nodes, or 4 nodes and it will work **Kurtosis environment definitions are completely reproducible and fully parametrizable.**

Reviewer: ```suggestion
Congratulations! You’ve just parameterized your Cassandra cluster environment definition and used it to instantiate a 5-node Cassandra cluster. You can run the same command with 2 nodes, or 4 nodes, and it will just work. **Kurtosis environment definitions are completely reproducible and fully parametrizable.**
```

:::caution

> Depending on how many nodes you wish to spin up, the max heap size of each node may cumulatively consume more memory on your local machine than you have available, causing the Starlark script to time-out. Modifying the `MAX_HEAP_SIZE` property may help, depending on your needs.

Reviewer: @leeederek  can you specify where the user would modify this property? just so its abundantly clear / user doesn't have to think

Author: Yup, done!

:::

### Review
How did that work?

> The plan object contains all enclave-modifying methods like `add_service`, `remove_service`, `upload_files`, etc. To accept arguments in the `run` function, we simply needed to pass them (arguments) in as the second parameter of our `run` function, like so:

Reviewer: ```suggestion
The plan object contains all enclave-modifying methods like `add_service`, `remove_service`, `upload_files`, etc. To use arguments, you accessed them via the second parameter of the `run` function, like so:
```

```python
def run(plan, args):
	...
```

> …which is what we used in our `main.star` file originally!

Reviewer: ```suggestion
…which is what you used in your `main.star` file originally!
```

> What we did next was add an `if statement` with the `hasattr()` function to tell Kurtosis that if an argument is passed in at runtime by a user, then override the default `num_nodes` variable, which we set as 3, with the user-specified value.

Reviewer: ```suggestion
What you did next was add an `if statement` with the `hasattr()` function to tell Kurtosis that if an argument is passed in at runtime by a user, then override the default `num_nodes` variable, which we set as 3, with the user-specified value.
```

> In our case, we passed in:

Reviewer: ```suggestion
In our case, you passed in:
```

```bash
'{"num_nodes": 5}'
```

> Which told Kurtosis to run your `main.star` file with 5 nodes instead of the default of 3 that we began with. 

Reviewer: ```suggestion
which told Kurtosis to run your `main.star` file with 5 nodes instead of the default of 3 that we began with. 
```

:::tip
Note that to pass parameters to the run(plan,args) function, a JSON object needs to be passed in as the 2nd position argument after the script or package path:
```bash
kurtosis run my_package.star '{"arg": "my_name"}'
```
:::

> In this section, we showed how easy it is to use Kurtosis to parameterize our environment definition with just a few lines of code. Next, we’ll walk through another powerful property of Kurtosis environments: composability.

Reviewer: ```suggestion
In this section, we showed how to use Kurtosis to parameterize an environment definition. Next, we’ll walk through another property of Kurtosis environments: composability.
```

> Using and making composable environment definitions

Reviewer: ```suggestion
Making and using composable environment definitions
```

---------------------------------------------------

> We’ve now written an environment definition using Starlark to instantiate a 3-node Cassandra cluster over Docker, and then modified it slightly to parameterize that definition for other use cases (n-node Cassandra cluster).

Reviewer: ```suggestion
You’ve now written an environment definition using Starlark to instantiate a 3-node Cassandra cluster over Docker, and then modified it slightly to parameterize that definition for other use cases (n-node Cassandra cluster).
```

> However, the benefits of parametrized and reproducible environment properties are amplified when we’re able to share our definition with others to use or use definitions that others (friends, colleagues, etc) have already written.

Reviewer: ```suggestion
However, the benefits of parametrized and reproducible environment properties are amplified when you’re able to share your definition with others to use, or when you use definitions that others (friends, colleagues, etc) have already written.
```

> To quickly demonstrate the latter capability, simply run:

Reviewer: ```suggestion
To quickly demonstrate the latter capability, run:
```

```bash
kurtosis clean -a && kurtosis run --enclave cassandra-cluster github.com/kurtosis-tech/cassandra-package '{"num_nodes": 5}' 
```

> Which should give you the same result as we saw when we ran our local `main.star` file with 5 nodes specified as an argument. However this time, we’re actually using a `main.star` file hosted [remotely on Github and running that instead][github-cass-package]!

Reviewer: ```suggestion
which should give you the same result you got when you ran your local `main.star` file with 5 nodes specified as an argument. However this time, you’re actually using a `main.star` file hosted [remotely on Github][github-cass-package]!
```

> **Any Kurtosis environment definition can be converted into a runnable, shareable artifact that we call a Kurtosis Package.** What we just did was run a remotely-hosted Kurtosis package.

Reviewer: ```suggestion
**Any Kurtosis environment definition can be converted into a runnable, shareable artifact that we call a Kurtosis Package.** 
```

> While this guide won’t cover the steps needed to convert your Starlark file and export it for others to use as a Kurtosis Package, you can check out our docs to learn more about how to turn a Starlark file into a [runnable Kurtosis package][runnable-packages] using only a [`kurtosis.yml`][kurtosis-yml] file.

Reviewer: ```suggestion
While this guide won’t cover the steps needed to convert your Starlark file and export it for others to use as a Kurtosis Package, you can check out our docs to learn more about how to turn a Starlark file into a [runnable Kurtosis package][runnable-packages].
```

### Review

> What we did with our most recent command was execute a remotely-hosted `main.star` from a Kurtosis package on [Github][github-cass-package]. That remotely-hosted `main.star` file has the same code as our local `main.star` file and, with only a [locator][locator], Kurtosis knew we were referencing a publicly-hosted runnable package. 

Reviewer: ```suggestion
You just executed a remote-hosted `main.star` from a Kurtosis package on [Github][github-cass-package]. That remote-hosted `main.star` file has the same code as our local `main.star` file and, with only a [locator][locator], Kurtosis knew that you were referencing a publicly-hosted runnable package. 
```

> The entirety of what we wrote locally in our `main.star` file can be packaged up and pushed to the internet (Github) for anyone to use to instantiate an n-node cassandra cluster by simply adding a [`kurtosis.yml`][kurtosis-yml] manifest to your directory and publishing the entire directory to Github. Read more about how to do this [here][runnable-packages].

Reviewer: ```suggestion
The entirety of what you wrote locally in your `main.star` file can be packaged up and pushed to the internet (Github) for anyone to use to instantiate an n-node cassandra cluster by adding a [`kurtosis.yml`][kurtosis-yml] manifest to your directory and publishing the entire directory to Github. Read more about how to do this [here][runnable-packages].
```

> Turning your Starlark code into a runnable Kurtosis package and making it available on Github means any developer will be able to use our Kurtosis package as a building block in their own environment definition or to run different tests using various configurations of nodes in a Cassandra cluster.

Reviewer: ```suggestion
Turning your Starlark code into a runnable Kurtosis package and making it available on Github means any developer will be able to use your Kurtosis package as a building block in their own environment definition or to run different tests using various configurations of nodes in a Cassandra cluster.
```

> This behavior is bi-directional as well. Meaning, we can import any remotely hosted Kurtosis package and use it’s code inside our own Kurtosis package with the `import_module` instruction like so:

Reviewer: ```suggestion
This behavior is bi-directional as well. Meaning, you can import any remotely hosted Kurtosis package and use its code inside your own Kurtosis package with the `import_module` instruction like so:
```

```python
lib = import_module("github.com/foo/bar/src/lib.star")

def run(plan,args)
lib.some_function()
lib.some_variable()
```

> **This feature we just witnessed is the composability property of environments created using Kurtosis and allows environments, and their components, to be easily composed and connected together without needing to know the inner workings of each setup.**

Reviewer: ```suggestion
**Package distribution via remote-hosted git repositories is one way in which Kurtosis enables environments to be easily composed and connected together without end users needing to know the inner workings of each setup.**
```

Conclusion
----------

> And that’s it! To recap this short guide, together we:

Reviewer: ```suggestion
And that’s it! To recap this short guide, you:
```

1. Wrote an environment definition that instructs Kurtosis to set up a 3 node Cassandra cluster in an isolated environment called an Enclave (over Docker),

> 2. We added a few lines of code to our `main.star` file to parametrize our environment definition to increase the flexibility and use cases with which it can be used, and,

Reviewer: ```suggestion
2. Added a few lines of code to our `main.star` file to parametrize your environment definition to increase the flexibility and use cases with which it can be used, and,
```

> 3. Executed a remotely-hosted Starlark file that was part of what's called a [Kurtosis Package][packages] to demonstrate the easy composability and reusability of Kurtosis environment definitions.

Reviewer: ```suggestion
3. Executed a remotely-hosted Starlark file that was part of what's called a [Kurtosis Package][packages] to demonstrate how your environment definition can be shared with other developers.
```

We’d love to hear from you on what went well for you, what could be improved, or to answer any of your questions. Don’t hesitate to reach out via Github (`kurtosis feedback`) or [email us](mailto:feedback@kurtosistech.com)!

### Other exercises you can do with your Cassandra cluster

> With our parameterized, reusable environment definition for a multi-node Cassandra cluster, feel free to:

Reviewer: ```suggestion
With your parameterized, reusable environment definition for a multi-node Cassandra cluster, feel free to:
```

* Turn your local `main.star` file into a runnable Kurtosis package and upload it on Github for others to use following [these instructions][runnable-packages]
* Use our [Go or Typescript SDK][sdk] to write backend integration tests, like this [network partitioning test][network-partitioning-test]

### Other examples

> While this was a short intro to some of Kurtosis’ capabilities, we encourage you to check out our [quickstart][quickstart] (where you’ll build a postgres database and API on top) and our other examples in our [awesome-kurtosis repository][awesome-kurtosis] where you will find other Kurtosis packages for you to check out as well, including a package that spins up a local [Ethereum testnet][eth-package-example] or one that sets up a [voting app using a Redis cluster][redis-package-example]. 

Reviewer: ```suggestion
We encourage you to check out our [quickstart][quickstart] (where you’ll build a postgres database and API on top) and our other examples in our [awesome-kurtosis repository][awesome-kurtosis] where you will find other Kurtosis packages for you to check out as well, including a package that spins up a local [Ethereum testnet][eth-package-example] or one that sets up a [voting app using a Redis cluster][redis-package-example]. 
```
